doc/design/attribute.md

```diff
@@ -0,0 +1,112 @@
+# Design Doc: Operator Attributes
+
+## Background
+
+An operator could have attributes. For example, CosineOp could have a float typed attribute scale, which changes the output range from [-1,1] to [-scale,scale]. The default value of scale is `1.0`.
+
+Attributes is defined by a name and a type. An instance of an attribute has a value of that type.
+
+As part of the network description, attribute need to be serialized. So we need a protobuf message that describes an attribute, say `Attribute`.
+
+An operator could parse the Attribute and save them into its private data member.
+
+## Protobuf Implementation
+
+There are two frameworks implement `Attribute` concept in `protobuf`. They are [`caffe2`](https://github.com/caffe2/caffe2/blob/master/caffe2/proto/caffe2.proto#L98) and [`tensorflow`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/attr_value.proto#L16).
+
+* Caffe2 uses `proto2` syntax. It treats all attributes as a list, and each attribute contains a `name`. Each time caffe2 read an attribute is searching a variable in a list. It is slow if the number of attributes is large. Caffe2 also mark all field as `optional`. It doesn't ensure `one of` attribute value is set.
+* By using `proto3` syntax in tensorflow, the attribute implementation in tensorflow is using `map`, and `oneof` keywords. Looking up from attribute map in tensorflow is fast.
+
+Paddle is using `protobuf 3` as its dependency library. By simplify `tensorflow`'s implementation, Paddle's Attribute protobuf message schema could be
+
+```protobuf
+message Attribute {
+    message ListValue {
+        repeated int32 ints = 1;
+        repeated float floats = 2;
+        repeated string strings = 3;
+    }
+
+    oneof value {
+        ListValue list = 1;
```

Comment:
The oneof directive doesn't seem help much here -- it can locate a field in `value` but it cannot help us locate a field in `ListValue`.  Our C++ code would still have to check which field is the one in `ListValue`.  If so, it would be easier just use the old plain syntax:

```protobuf
syntax = "proto3";

message Attribute {
    repeated int32 ints = 1;
    repeated float floats = 2;
    repeated string strings = 3;
    int32 int = 4;
    float float = 5;
    string string = 6;
}
```